font-types/src/pen.rs

```diff
@@ -85,3 +91,146 @@ impl PenCommand {
         }
     }
 }
+
+/// Pen adapter that outputs commands to SVG path data.
+///
+/// The target may be any type that implements `fmt::Write` such as
+/// `String`.
+pub struct SvgPen<T> {
+    target: T,
+    space: &'static str,
+}
+
+impl<T> SvgPen<T> {
+    pub fn new(target: T) -> Self {
+        Self { target, space: "" }
+    }
+
+    pub fn into_inner(self) -> T {
+        self.target
+    }
+}
+
+impl<T: fmt::Write> Pen for SvgPen<T> {
+    fn move_to(&mut self, x: f32, y: f32) {
+        let _ = write!(self.target, "{}M{},{}", self.space, x, y);
+        self.space = " ";
+    }
+
+    fn line_to(&mut self, x: f32, y: f32) {
+        let _ = write!(self.target, "{}L{},{}", self.space, x, y);
+        self.space = " ";
+    }
+
+    fn quad_to(&mut self, cx0: f32, cy0: f32, x: f32, y: f32) {
+        let _ = write!(self.target, "{}Q{},{} {},{}", self.space, cx0, cy0, x, y);
+        self.space = " ";
+    }
+
+    fn curve_to(&mut self, cx0: f32, cy0: f32, cx1: f32, cy1: f32, x: f32, y: f32) {
+        let _ = write!(
+            self.target,
+            "{}C{},{} {},{} {},{}",
+            self.space, cx0, cy0, cx1, cy1, x, y
+        );
+        self.space = " ";
+    }
+
+    fn close(&mut self) {
+        let _ = write!(self.target, "{}Z", self.space);
+        self.space = " ";
+    }
+}
+
+/// Pen adapter that collects commands into a target buffer.
+///
+/// The target may be any type that implements `Extend<PenCommand>` such as
+/// `Vec<PenCommand>`.
+pub struct BufferPen<T> {
+    target: T,
+}
+
+impl<T> BufferPen<T> {
+    pub fn new(target: T) -> Self {
+        Self { target }
+    }
+
+    pub fn into_inner(self) -> T {
+        self.target
+    }
+}
+
+impl<T: Extend<PenCommand>> Pen for BufferPen<T> {
```

Comment:
I think this type bound makes it possible to create a BufferPen that has unexpected behaviour?

for instance I'm not sure if `PenCommand` implements `Eq` and `Ord`, but if so, if you used a `BTreeSet` as the internal storage you could end up storing commands in a different order than they were provided, or you could end up removing duplicate commands that were meaningful.

Reply:
Fair enough! I don’t have any concrete use cases at the moment beyond providing some actual useful impls of Pen that are available out of the box. The skrifa scaler docs would benefit from not having to define an impl inline to demonstrate usage but I don’t think that is a strong enough case to override your concerns.

The other motivation was trying to expose some of the pen types that are hidden away in write-fonts, but pursuing iterator based path transforms is probably a more interesting route in the long term.

I’ll leave this open for a bit and let it simmer but based on your comments, I’m leaning towards closing and revisiting in the future.

Reply:
> I wonder if the use of generics is maybe solving a problem we don't have?

+1, this looks overengineered at a glance.

Reply:
> this seems to already exist in write-fonts (backed by a Vec)

We should delete them?